# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: third_party/rust/rental/tests/lt_params.rs
# Commit: 16350e08d34f
# Full Hash: 16350e08d34ff06c7f46745c6b160c7a7bda9140
# Author: Dana Keeler <dkeeler@mozilla.com>
# Date: 2019-12-06 04:43:57
# Regressor Bug: 1586855
# File Overlap Count: 6
# Description:
#   bug 1586855 - incorporate CRLite filters into cert_storage r=jcj,kjacobs
#   
#   This patch implements CRLite lookups for TLS server certificate revocation
#   information in telemetry-only mode. It adds a new preference
#   "security.pki.crlite_mode" to control the behavior of this feature. Setting
# ==============================================================================

diff -r 64fd40663930 -r 16350e08d34f third_party/rust/rental/tests/lt_params.rs
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/third_party/rust/rental/tests/lt_params.rs	Thu Dec 05 22:41:53 2019 +0000
@@ -0,0 +1,65 @@
+#[macro_use]
+extern crate rental;
+
+
+pub struct Foo<'a> {
+	i: &'a i32,
+}
+
+impl<'a> Foo<'a> {
+	fn borrow(&self) -> &i32 { self.i }
+	fn try_borrow(&self) -> Result<&i32, ()> { Ok(self.i) }
+	fn fail_borrow(&self) -> Result<&i32, ()> { Err(()) }
+}
+
+
+rental! {
+	mod rentals {
+		use super::*;
+
+		#[rental]
+		pub struct LtParam<'a> {
+			foo: Box<Foo<'a>>,
+			iref: &'foo i32,
+		}
+	}
+}
+
+
+#[test]
+fn new() {
+	let i = 5;
+
+	let foo = Foo { i: &i };
+	let _ = rentals::LtParam::new(Box::new(foo), |foo| foo.borrow());
+
+	let foo = Foo { i: &i };
+	let sr = rentals::LtParam::try_new(Box::new(foo), |foo| foo.try_borrow());
+	assert!(sr.is_ok());
+
+	let foo = Foo { i: &i };
+	let sr = rentals::LtParam::try_new(Box::new(foo), |foo| foo.fail_borrow());
+	assert!(sr.is_err());
+}
+
+
+#[test]
+fn read() {
+	let i = 5;
+
+	let foo = Foo { i: &i };
+
+	let mut sr = rentals::LtParam::new(Box::new(foo), |foo| foo.borrow());
+
+	{
+		let i: i32 = sr.rent(|iref| **iref);
+		assert_eq!(i, 5);
+	}
+
+	{
+		let iref: &i32 = sr.ref_rent(|iref| *iref);
+		assert_eq!(*iref, 5);
+	}
+
+	assert_eq!(sr.rent_all_mut(|borrows| *borrows.foo.i), 5);
+}